### DISCUSSION TITLE
find midi tempo of a file

### DISCUSSION THREAD

**Discussion by csarami** [mido/mido#430] [Q&A] [answered]
Hi,

Is there a method to find the tempo of a midi file? Now, I go to the header chuck and find meta-message for that.

**Comment by olemb** [comment, marked as answer]
A MIDI file has a default tempo of 500000 microseconds per tick (120 beats per minute). The tempo can be changed at any point in the file with a `set_tempo` message, for example:

```python
MetaMessage('set_tempo', tempo=250000)
```

You can look for tempo messages with something like this:

```python
mid = MidiFile('test.mid')
for track in mid:
    for msg in track:
        if msg.type == 'set_tempo':
            print('Tempo set to', msg.tempo)
```

If the whole file is in one tempo there's usually a `set_tempo` at the start of one of the tracks, and you can find the tempo with something like this:

```python
def get_tempo(mid):
    for track in mid.tracks:
        for msg in track:
            if msg.type == 'set_tempo':
                return msg.tempo
    else:
        # Default tempo.
        return 500000
```

For more on tempo: https://mido.readthedocs.io/en/latest/midi_files.html#tempo-and-beat-resolution

**Comment by EFeru** [comment]
@olemb thanks for the input. However if I do your last suggestion I get the following error:
```python
>>> get_tempo(mid)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in get_tempo
TypeError: 'MetaMessage' object is not iterable
```
Thus, I updated your script to:
```python
def get_tempo(mid):
    for msg in mid:     # Search for tempo
        if msg.type == 'set_tempo':
            return msg.tempo
    return 500000       # If not found return default tempo
```

Then I get the correct result for my midi file:
```python
>>> get_tempo(mid)
909091
```

---
**Edit:** The error in your script is `for track in mid:` should be `for track in mid.tracks:`. Then your script work too:
```python
>>> def get_tempo(mid):
...     for track in mid.tracks:
...         for msg in track:
...             if msg.type == 'set_tempo':
...                 return msg.tempo
...     else:
...         # Default tempo.
...         return 500000
...
>>> get_tempo(mid)
909091
```
